text/0030-shared-event-queue.md

```diff
@@ -0,0 +1,122 @@
+[performance optimization][0] for the case where a single `ThreadsafeFunction`
+is invoked multiple times in a row. Thus using a single shared
+`ThreadsafeFunction` in `EventQueue`s would lead to significant performance
+improvements for Neon users.
+
+These improvements are perhaps most noticeable when Neon addon is used from the
+Electron, because scheduling new UV ticks in Electron are very costly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Whenever `cx.queue()` is invoked - a new instance of `EventQueue` is created.
+With ithe proposed changes the instance would point to a shared
+`ThreadsafeFunction` internally. For the most scenarios there would be no
+difference between current and proposed implementation. Calling `queue.send()`
+will still invoke the closure on the JavaScript thread:
+```rust
+let queue = cx.queue();
+
+std::thread::spawn(move || {
+    queue.send(|| {
+        Ok(())
+    })
+});
+```
+
+Calling `queue.send()` several times in a row using different `cx.queue()` would
+be significantly faster:
+```rust
+loop {
+    cx.queue().send(|| {
+        Ok(())
+    })
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+With the recent [node.js changes][0] `ThreadsafeFunction` will loop over pending
+calls, invoking up to 1000 of them in the same tick instead of running each new
+call on a separate UV loop tick. This is an implementation detail that is
+intentionally not covered by `napi_create_threadsafe_function` documentation to
+allow for potential performance optimizations in the future.
+
+Neon doesn't currently leverage it without the proposed changes because a new
+`ThreadsafeFunction` is created for each new `EventQueue`, and `cx.queue()`
+creates a new `EventQueue` every time as requested. So instead of doing that it
+is proposed that we use existing `lifetime::InstanceData` internal API layer to
+hold a wrapper around `ThreadsafeFunction` tailored to execute Rust closures
+specifically, and to use an `Arc` reference to a `RwLock` over this wrapper in
+the `EventQueue` for a shared access.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If some of the users of `cx.queue()` would post significantly more closures on
+the queue than the others it might lead to a potential starvation:
+```rust
+let queue = cx.queue();
+std::thread::spawn(move || {
+    loop {
+        thread::sleep(Duration::from_secs(1));
+
+        queue.send(|| {
+            println!("every second");
+            Ok(())
+        })
+    }
+});
+
+let another_queue = cx.queue();
+std::thread::spawn(move || {
+    loop {
+        another_queue.send(|| {
+            println!("immediate");
+            Ok(())
+        })
+    }
+});
+```
+
+The example above wouldn't print "every second" once per second. The prints
+will be delayed by the closures scheduled on `another_queue`.
+
+While the example above reproduces the issue - it is quite pathological and
```

Comment:
I think this use case is valid. The example is a bit extreme for the sake of simplicity to demonstrate the point, but I do think it could be an issue in real apps.

For example, consider a Node app that's part of a data processing pipeline. It would be _expected_ to saturate resources and that might include a stream of events on the event loop. It may be beneficial to also be able to _cancel_ the pipeline at any time.

If command events are served on a separate channel from processing events, it allows the app to push large volumes of data without delaying commands. Alternatively, the app would need to manage back-pressure.

Reply:
The data processing pipeline would likely be bound by IO and so it wouldn't be able to trigger the 1000 call loop in the n-api optimization, but the starvation can indeed be triggered by posting way too many events that take long time to process from one producer and some events from another producer that should interrupt that stream.

Was the preemption ever documented for neon APIs? I'm looking through the comments in the EventQueue.rs and the only thing that they say is that the function will be executed in JS thread without saying anything about preemption between different queues.